Post: You wanna test HTTP, right? (#232)
mauroservienti/milestone.github.io: _posts/you-wanna-test-http.md

---
layout: post
header_image: /img/posts/you-wanna-test-http/header.jpg
title: "You wanna test HTTP, right?"
author: Mauro Servienti
synopsis: "Tests generate confidence. Awesome. Sometimes we need to test components requiring HTTP, and we risk building complex test infrastructures that only erode confidence. What options do we have?"
tags:
- unit test
- integration test
- HTTP
---

One of the beauties of testing is that it generates confidence. It's genuinely the essential aspect. Code evolves, needs to be maintained and fixed. We want to be confident that a set of changes will not change the expected behavior; we need to be sure we're not reintroducing bugs or undesirable behaviors that we already fixed.

There is a need for safety, and testing creates that.

> Testing practices alone are not enough to create the required safety. Manual testing, for example, is a form of testing that comes with little sense of security: manual testing is prone to human and distraction errors. To build the required level of trust, we need attributes such as repeatability, isolation, and automation. Tests need to be automated so that they can run automatically at well-known predefined stages. Tests need to be repeatable and idempotent to guarantee consistency of results; Finally, tests need to be isolated and self-contained to ensure that one test doesn't affect others.

Reviewer: ```suggestion
But testing practices alone are not enough to create the required safety. Manual testing, for example, is a form of testing that comes with little sense of security: manual testing is prone to human and distraction errors. To build the required level of trust, we need attributes such as repeatability, isolation, and automation. Tests need to be automated so that they can run automatically at well-known predefined stages. Tests need to be repeatable and idempotent to guarantee consistency of results. Finally, tests need to be isolated and self-contained to ensure that one test doesn't affect others.
```

Isolation comes with some headaches. Typically we start facing troubles when the tested components depend on external resources. For example, if we're testing a repository component, we need a data source to test that the repository does what it promises. The dependency on external resources generally surfaces as a problem the more we approach the boundaries of the system under test. A business component might depend on the repository mentioned above; in a test, we can mock the repository and pilot the business component tests. When testing the repository, it's hard or even impossible to mimic the required data source.

> HTTP resources pose a similar issue. Let's imagine that you want to use [approval testing](https://approvaltests.com) techniques to validate that an HTTP endpoint returns the expected result. The test exercises the client of the HTTP resource, that is, the system under test. In this scenario, the HTTP resource is not different from the database required by the repository mentioned above. As for databases, we cannot mock HTTP resources. At the same time, hosting a webserver in a test is not straightforward.

Reviewer: ```suggestion
HTTP resources pose a similar issue. Let's imagine that you want to use [approval testing](https://approvaltests.com) techniques to validate that an HTTP endpoint returns the expected result. The test exercises the client of the HTTP resource, that is, the system under test. In this scenario, the HTTP resource is not different from the database required by the repository mentioned above. But unlike databases, we cannot mock HTTP resources. At the same time, hosting a webserver in a test is not straightforward.
```

   private readonly RequestDelegate _next;

   public RequestCultureMiddleware(RequestDelegate next)
   {
      _next = next;
   }

   public async Task InvokeAsync(HttpContext context)
   {
      var cultureQuery = context.Request.Query["culture"];
      if (!string.IsNullOrWhiteSpace(cultureQuery))
      {
         var culture = new CultureInfo(cultureQuery);
         CultureInfo.CurrentCulture = culture;
         CultureInfo.CurrentUICulture = culture;
      }

      _next(context);
   }
}
```

> The complete sample is available on the [official Microsoft documentation.](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/middleware/write?view=aspnetcore-5.0)

To use the custom middleware, we'd generally write an extension method like the following:

```csharp
public static class RequestCultureMiddlewareExtensions
{
   public static IApplicationBuilder UseRequestCulture( this IApplicationBuilder builder)
   {
      return builder.UseMiddleware<RequestCultureMiddleware>();
   }
}
```

That allows to configure the web application in the following way:

```csharp
public void Configure(IApplicationBuilder app)
{
   app.UseRequestCulture();
   //more configuration here
}
```

> If we wanted to test the middleware, we could use the `WebApplicationFactory<TEntryPoint>` mentioned above. Still, we should include a dummy web application project in the solution just to referencing the `Startup` class. That sounds overkill.

Reviewer: ```suggestion
If we wanted to test the middleware, we could use the `WebApplicationFactory<TEntryPoint>` class mentioned above. Still, we should include a dummy web application project in the solution just to reference the `Startup` class. That sounds like overkill.
```

> Carefully looking at the [official documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1?view=aspnetcore-5.0), it sounds that's not needed at all:

Reviewer: ```suggestion
Carefully looking at the [official documentation](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.testing.webapplicationfactory-1?view=aspnetcore-5.0), it sounds like that's not needed at all:
```

        hostBuilder.ConfigureWebHostDefaults(webBuilder =>
        {
            webBuilder.ConfigureServices(configureServices);
            webBuilder.Configure(configure);

            WebHostBuilderCustomization?.Invoke(webBuilder);
        });

        HostBuilderCustomization?.Invoke(hostBuilder);

        return hostBuilder;
    }
}
```

I feel that what the above class is doing is pretty straightforward and doesn't need any comment. The presented `WebApplicationFactoryWithHost` is nothing more than a sample. There are many enhancements that we can apply. It just works for my use case.

Let's have a look at the usage:

```csharp
[Fact]
async Task SampleMiddlewareTest()
{
   var factory = new WebApplicationFactoryWithHost<Dummy>
   (
      configureServices: services =>
      {
         //configure services as needed
      }
      configure: app =>
      {
         app.UseRequestCulture();
         //rest of the required app configuration here
      }
   );

   var client = factory.CreateClient();
   var response = await client.GetAsync("/sample/1?culture=it-IT");

   Assert.True(response.IsSuccessStatusCode);
   //more assertions to validate the culture settings
}
```

> Cool, isn't it? We're self-hosting the whole web application in the test. We control the services customization and the application set up when started. Finally, we can exercise it like if it was an actual web application.

Reviewer: ```suggestion
Cool, isn't it? We're self-hosting the whole web application in the test. We control the services customization and the application set up when started. Finally, we can exercise it like it was an actual web application.
```

> > The `Dummy` class is nothing else that a marker/empty class defined in the test project. It can be whatever type in the test project; the runtime will use it to identify the assembly to create the dependency context to load dependencies.

Reviewer: ```suggestion
> The `Dummy` class is simply a marker/empty class defined in the test project. It can be whatever type in the test project; the runtime will use it to identify the assembly to create the dependency context to load dependencies.
```
